File: 24-s3-presigned-urls-with-lambda/lambda-function-code.py
import os
import boto3
import json
from botocore.exceptions import ClientError

s3_client = boto3.client('s3')
BUCKET_NAME = os.environ['BUCKET_NAME']

def lambda_handler(event, context):
    # Parse object key and target operation from the incoming request body
    body = json.loads(event.get('body', '{}'))
    object_key = body.get('key')
    operation = body.get('operation', 'get_object') # Defaults to download ('get_object')
    
    if not object_key:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Missing file "key" parameter'})
        }

    try:
        # Generate the presigned URL valid for 15 minutes (900 seconds)
        response = s3_client.generate_presigned_url(
            ClientMethod=operation, # 'get_object' for download or 'put_object' for upload
            Params={'Bucket': BUCKET_NAME, 'Key': object_key},
            ExpiresIn=900 
        )
    except ClientError as e:
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }

    return {
        'statusCode': 200,
        'body': json.dumps({'presigned_url': response})
    }

File: 24-s3-presigned-urls-with-lambda/lambda_function_code.js
import { S3Client, GetObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

export const handler = async (event) => {
    const s3Client = new S3Client({ region: "us-east-1" });
    
    const bucket_name = "private-bucket-49587sdf90458s3";
    
    // this must be the name of the file to upload and download
    const object_key = "example-file.txt";

    try {
        // Create the command (use PutObjectCommand for file uploads)
        const command = new GetObjectCommand({
            Bucket: bucket_name,
            Key: object_key,
        });

        // Generate the presigned URL
        const presignedUrl = await getSignedUrl(s3Client, command, { 
            expiresIn: 300 // URL expires in 5 minutes (300 seconds)
        });

        return {
            statusCode: 200,
            headers: {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*"
            },
            body: JSON.stringify({ url: presignedUrl }),
        };
    } catch (error) {
        console.error("Error creating presigned URL", error);
        return {
            statusCode: 500,
            body: JSON.stringify({ error: "Failed to generate presigned URL" }),
        };
    }
};
